brainstorming.md

```diff
@@ -0,0 +1,691 @@
+
+### SOMABatchSize
+
+Read operations on foundational types return an iterator over "batches" of data, enabling the processing of larger-than-core datasets. The SOMABatchSize allows user control over read batch size, and accepts the following methods of determining batch size:
+
+| BatchSize type | Description                                                                                                                                         |
+| -------------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `count`        | Batch size defined by result count. For a SOMADataFrame this indicates row count returned per RecordBatch, or for an ND array the number of values. |
+| `size`         | Partition defined by size, in bytes, e.g., max RecordBatch size returned by SOMADataFRame read operation.                                           |
+| `auto`         | An automatically determined, "reasonable" default partition size. This is the default batch size.                                                   |
+
+### SOMAReadPartitions
+
+To facilitate distributed computation, read operations on foundational types accept a user-specified parameter indicating the desired partitioning of reads and which partition any given read should return. The following options are supported:
+
+| Partition Type | Description                                                                                                                                                                        |
+| -------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `IofN`         | Given I and N, read operations will return the Ith partition of N approximately equal size partitions. Partition boundaries will be stable for any given N and array or dataframe. |
+
+### SOMABatchFormat
+
+Array read operations can return results in a variety of formats. The `SOMABatchFormat` format indicates the format encoding.
+
+| Batch format   | Description                                                                                           |
+| -------------- | ----------------------------------------------------------------------------------------------------- |
+| `dense`        | Return the coordinates of the slice (e.g. origin, shape) and an Arrow Tensor containing slice values. |
+| `coo`          | Return an Arrow.SparseCOOTensor                                                                       |
+| `csr`          | Return an Arrow.SparseCSRTensor                                                                       |
+| `csc`          | Return an Arrow.SparseCSCTensor                                                                       |
```

Comment:
Just wanted to drop in on this:

As I've mentioned before (https://github.com/single-cell-data/SOMA/issues/18#issuecomment-1075398109) I'm a little concerned about arrow's level of support for its Tensor types, especially the sparse tensor types. Last I looked into this, it seemed like there were no tests in the arrow ipc suite for these. E.g. there's no confirmation interchange works for tensors. I believe this has led to a number of implementations just skipping out on tensor (esp. sparse tensor) support.

That said, I would really love it if these were better supported.

Reply:
I hear ya...that said, I am unsure there is any good (portable, common, etc) alternative other than removing the capability (ie, only supporting COO in a RecordBatch) _or_ defining our own data structure. IMHO, both are worse than taking the risk of being an early SparseTensor adoptor.

If I'm missing an alternative, would love to hear your ideas!

Reply:
The only alternative I can think of is how vitessce handles sparse zarr AnnData matrices.

In effect, I'm not sure how different this is than defining a custom data structure, since clients will have to handle decoding.

My ideal here would be that there is some contact with the arrow team, there's some work on what they think about it, and some ownership of these structures goes to people who care about them. Tests could get added to the IPC, so implementation start supporting these (for example).

Following from SciPy this year, some discussions have started around binary formats for sparse data. Discussions are being recorded here: https://github.com/GraphBLAS/binsparse-specification and is largely w/ scientific-python and graphBLAS. Still quite early though.